src/Analyzers/CSharp/Analyzers/UsePatternCombinators/CSharpUsePatternCombinatorsDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,144 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System.Linq;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Shared.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.CSharp.UsePatternCombinators
+{
+    using static AnalyzedPattern;
+
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    internal sealed class CSharpUsePatternCombinatorsDiagnosticAnalyzer :
+        AbstractBuiltInCodeStyleDiagnosticAnalyzer
+    {
+        public CSharpUsePatternCombinatorsDiagnosticAnalyzer()
+            : base(IDEDiagnosticIds.UsePatternCombinatorsDiagnosticId,
+                CSharpCodeStyleOptions.PreferPatternMatching,
+                LanguageNames.CSharp,
+                new LocalizableResourceString(nameof(CSharpAnalyzersResources.Use_pattern_matching), CSharpAnalyzersResources.ResourceManager, typeof(CSharpAnalyzersResources)),
+                new LocalizableResourceString(nameof(CSharpAnalyzersResources.Use_pattern_matching), CSharpAnalyzersResources.ResourceManager, typeof(CSharpAnalyzersResources)))
+        {
+        }
+
+        protected override void InitializeWorker(AnalysisContext context)
+            => context.RegisterSyntaxNodeAction(AnalyzeNode,
+                SyntaxKind.SwitchExpressionArm,
+                SyntaxKind.ConditionalExpression,
+                SyntaxKind.ForStatement,
+                SyntaxKind.EqualsValueClause,
+                SyntaxKind.IfStatement,
+                SyntaxKind.WhenClause,
+                SyntaxKind.WhileStatement,
+                SyntaxKind.DoStatement,
+                SyntaxKind.ReturnStatement,
+                SyntaxKind.YieldReturnStatement,
+                SyntaxKind.ArrowExpressionClause,
+                SyntaxKind.SimpleAssignmentExpression,
+                SyntaxKind.SimpleLambdaExpression,
+                SyntaxKind.ParenthesizedLambdaExpression,
+                SyntaxKind.Argument);
```

Comment:
so i don't love this due to the need to hard code this different list.  could we not just look for certain expressions (binary, unary, parenthesized), and then only analyze them if they're the topmost expr?

Reply:
with this we'll get a lot less hits. if that's not significant perf-wise, I can look for exprs. however, even then we'll have to hard-code *another* list?

Reply:
sure.  but the latter list is very intuitive and in line with what's being analyzed.  i.e. we're looking to simplify `!` and `<` and `&&` expressions and the like.  So looking for those cases just makes more sense to me.